src/nvim/eval.c

```diff
@@ -8495,6 +8495,25 @@ static void f_empty(typval_T *argvars, typval_T *rettv, FunPtr fptr)
   rettv->vval.v_number = n;
 }
 
+/// "environ()" function
+static void f_environ(typval_T *argvars, typval_T *rettv, FunPtr fptr)
+{
+  tv_dict_alloc_ret(rettv);
+
+  for (int i = 0; ; i++) {
+    // TODO(justinmk): use os_copyfullenv from #7202 ?
+    char *envname = os_getenvname_at_index((size_t)i);
+    if (envname == NULL) {
+      break;
+    }
+    const char *value = os_getenv(envname);
+    tv_dict_add_str(rettv->vval.v_dict,
+                    (char *)envname, STRLEN((char *)envname),
+                    value == NULL ? "" : value);
```

Comment:
For the `value == NULL` case we should have a test like https://github.com/vim/vim/pull/4751/files - probably also as a busted test (assuming Vim is not taking that one also - or it would need some special handling there, e.g. setting it outside of Vim already).

Reply:
doesn't matter too much, whatever is fastest/easiest.